bsp/Src/I2C.c

```diff
@@ -0,0 +1,423 @@
+			}
+		}
+	}
+	return HAL_OK;
+}
+
+/**
+* @brief    Receives data from the I2C bus.
+* @param    hi2c:            to structure that has config data
+* @param    deviceAddr      target device address
+* @param    pDataBuff      data buffer
+* @param    len            amount of data
+* @return   HAL_StatusTypeDef
+*/
+HAL_StatusTypeDef i2c_recv(I2C_HandleTypeDef *hi2c,
+              uint8_t deviceAddr,
+              uint8_t* pDataBuff,
+              uint16_t len)
+{
+	// Update the latest I2C peripheral
+	update_last_i2c(hi2c);
+
+    return HAL_I2C_Master_Receive_IT(hi2c, deviceAddr, pDataBuff, len);
+}
+
+/**
+ * @brief 	Internal function to send the value of the specified i2c
+ * peripheral's queue
+ *
+ * @param	i2c_peripheral Either I2C1, I2C1, or I2C3
+ */
+static void queue_send(I2C_TypeDef *i2c_peripheral)
+{
+	I2C_HandleTypeDef *last_hi2c = NULL;
+	QueueHandle_t *I2C_Queue = NULL;
+	BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+	#ifdef I2C1
+	if (i2c_peripheral == I2C1)
+	{
+		last_hi2c = last_hi2c1;
+		I2C_Queue = &I2C1_Queue;
+	}
+	#endif
+	#ifdef I2C2
+	if (i2c_peripheral == I2C2)
+	{
+		last_hi2c = last_hi2c2;
+		I2C_Queue = &I2C2_Queue;
+	}
+	#endif
+	#ifdef I2C3
+	if (i2c_peripheral == I2C3)
+	{
+		last_hi2c = last_hi2c3;
+		I2C_Queue = &I2C3_Queue;
+	}
+	#endif
+
+	if(I2C_Queue == NULL || last_hi2c == NULL)
+	{
+		// Invalid I2C peripheral
+		return;
+	}
+
+	if (HAL_I2C_GetState(last_hi2c) != HAL_I2C_STATE_READY)
+		return;
+
+	DataInfo_t data;
+	if (xQueueReceiveFromISR(*I2C_Queue, &data, &higherPriorityTaskWoken) == pdPASS) {
+	    // Process dequeuedData
+		if(i2c_peripheral == I2C1)
+		{
+			uint8_t dataToTransmit[I2C1_MAX_PAYLOADS];
+			uint8_t index = 0;
+			dataToTransmit[index] = data.pDataBuff;
+			while(data.payloadsLeft != 0)
+			{
+				xQueueReceiveFromISR(*I2C_Queue, &data, &higherPriorityTaskWoken);
+				dataToTransmit[++index] = data.pDataBuff;
+			}
+
+			HAL_I2C_Master_Transmit_IT(data.hi2c, data.deviceAddr, dataToTransmit, index + 1);	
+		}
+		else if(i2c_peripheral == I2C2)
+		{
+			uint8_t dataToTransmit[I2C2_MAX_PAYLOADS];
+			uint8_t index = 0;
+			dataToTransmit[index] = data.pDataBuff;
+			while(data.payloadsLeft != 0)
+			{
+				xQueueReceiveFromISR(*I2C_Queue, &data, &higherPriorityTaskWoken);
+				dataToTransmit[++index] = data.pDataBuff;
+			}
+
+			HAL_I2C_Master_Transmit_IT(data.hi2c, data.deviceAddr, dataToTransmit, index + 1);	
+		}
+		else if(i2c_peripheral == I2C3)
+		{
+			uint8_t dataToTransmit[I2C3_MAX_PAYLOADS];
+			uint8_t index = 0;
+			dataToTransmit[index] = data.pDataBuff;
+			while(data.payloadsLeft != 0)
+			{
+				xQueueReceiveFromISR(*I2C_Queue, &data, &higherPriorityTaskWoken);
+				dataToTransmit[++index] = data.pDataBuff;
+			}
+
+			HAL_I2C_Master_Transmit_IT(data.hi2c, data.deviceAddr, dataToTransmit, index + 1);	
+		}
+		
+		// Control should not reach here as only valid I2C should be able to enter the queue
+	} else {
+	    // Handle queue empty error

```

Comment:
This is where error enums could be useful. At the very least to return what specific error happened.

Reply:
Removed this logic because the queue being empty is not an error. It simply means there is nothing more to transmit.